### PR TITLE
add `ResponseException#status`

### PR DESCRIPTION
**Pull request by rursprung**
this allows accessing the HTTP status code of the response when an API returns a `ResponseException`. this was not possible through `getResponse` since `Response` itself is package-private and thus its members (even though they are marked as `public`) are not accessible to external consumers.

as there's no test coverage for `ResponseException` i haven't added any for the new API either, as it is a trivial API and it was unclear how adding test coverage for it would fit into the existing test coverage.

solves #749

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

unrelated fun fact: 756 (the PR ID) is the ISO country code for my country 😄🇨🇭

### CODE REVIEW

**Comment by rursprung** [comment]
CC @reta & @dblock

File: java-client/src/main/java/org/opensearch/client/transport/httpclient5/ResponseException.java [hunk 1]
```diff
@@ -73,7 +73,7 @@ static String buildMessage(Response response) throws IOException {
 
         HttpEntity entity = response.getEntity();
         if (entity != null) {
-            if (entity.isRepeatable() == false) {
+            if (!entity.isRepeatable()) {
```
Comment (reta) [inline java-client/src/main/java/org/opensearch/client/transport/httpclient5/ResponseException.java:76]: Could you please revert this change? (this is the convention we use in OpenSearch for boolean checks, for better or worth)

File: java-client/src/main/java/org/opensearch/client/transport/httpclient5/ResponseException.java [hunk 2]
```diff
@@ -64,7 +64,7 @@ static String buildMessage(Response response) throws IOException {
             response.getRequestLine().getMethod(),
             response.getHost(),
             response.getRequestLine().getUri(),
-            response.getStatusLine().toString()
+            response.getStatusLine()
```
Comment (reta) [inline java-client/src/main/java/org/opensearch/client/transport/httpclient5/ResponseException.java:67]: Let's pass `this.response.getStatusLine().getStatusCode()`? We don't need the full `StatusLine` here ...

Comment (rursprung) [reply java-client/src/main/java/org/opensearch/client/transport/httpclient5/ResponseException.java:67]: hm, that was just an unrelated cleanup (the `toString` happens automatically) which my IDE nicely suggested - i can change it to `getStatusCode()`, but that'd change the behaviour - if you're ok with that i can do it (though i'll split it out into a separate commit so that it's more obvious)

File: java-client/src/main/java/org/opensearch/client/transport/httpclient5/ResponseException.java [hunk 3]
```diff
@@ -92,4 +92,12 @@ static String buildMessage(Response response) throws IOException {
     public Response getResponse() {
         return response;
     }
+
+    /**
+     * Status code returned by Elasticsearch. Shortcut for
```
Comment (reta) [inline java-client/src/main/java/org/opensearch/client/transport/httpclient5/ResponseException.java:97]:
```suggestion
     * Status code returned by Opensearch. Shortcut for
```

File: java-client/src/main/java/org/opensearch/client/transport/httpclient5/ResponseException.java (hunk 2 above)
Comment (reta) [reply java-client/src/main/java/org/opensearch/client/transport/httpclient5/ResponseException.java]: > hm, that was just an unrelated cleanup

Oh I see, sure, let's use the `response.getStatusLine()`, apologies for confusion

File: java-client/src/main/java/org/opensearch/client/transport/httpclient5/ResponseException.java (hunk 3 above)
Comment (dblock) [reply java-client/src/main/java/org/opensearch/client/transport/httpclient5/ResponseException.java:97]: capitalize OpenSearch accordingly.=

**Review by dblock — CHANGES_REQUESTED** [review]
Code looks good! Does need a test even if it's just to exercise the code to make sure it really works and we don't accidentally remove it.

**Comment by dblock** [comment]
> unrelated fun fact: 756 (the PR ID) is the ISO country code for my country 😄🇨🇭

Hopp Schwiiz!!

File: java-client/src/main/java/org/opensearch/client/transport/httpclient5/ResponseException.java [hunk 4]
```diff
@@ -92,4 +92,12 @@ static String buildMessage(Response response) throws IOException {
...
+     * {@code response().status()}.
+     */
+    public int status() {
```
Comment (dblock) [inline java-client/src/main/java/org/opensearch/client/transport/httpclient5/ResponseException.java:100]: Q: Should it be `status` or `getStatus`? What's the convention?

File: java-client/src/main/java/org/opensearch/client/transport/httpclient5/ResponseException.java (hunk 3 above)
Comment (rursprung) [reply java-client/src/main/java/org/opensearch/client/transport/httpclient5/ResponseException.java:97]: that was a copy & paste thing (it still says `Elasticsearch` on `OpenSearchException` where i copied this from and i hadn't seen it). i've checked and there are more such places, thus i've now raised #757 to get them all fixed (i'll fix it here as it's new code but won't touch `OpenSearchException` as part of this PR)

File: java-client/src/main/java/org/opensearch/client/transport/httpclient5/ResponseException.java (hunk 4 above)
Comment (rursprung) [reply java-client/src/main/java/org/opensearch/client/transport/httpclient5/ResponseException.java:100]: i took this 1:1 from `OpenSearchException`, presuming that this was the accepted convention. happy to change it to `getStatus`

Comment (rursprung) [reply java-client/src/main/java/org/opensearch/client/transport/httpclient5/ResponseException.java:99]: i've left it at `status` for now pending your feedback

**Comment by rursprung** [comment]
> Does need a test even if it's just to exercise the code to make sure it really works and we don't accidentally remove it.

ok, i've added a small test for it (though i skipped on testing the existing things as `buildMessage` is just creating a human-readable string, so parsing it in a test is IMHO adding little benefit and `getResponse` is package-private)

**Review by reta — APPROVED** [review]

**Comment by rursprung** [comment]
the test failure is a false reject due to infrastructure issues:
```
Downloading https://services.gradle.org/distributions/gradle-8.4-bin.zip

Error: Exception in thread "main" java.net.SocketException: Software caused connection abort: socket write error
```

can you re-trigger it? (i don't have the rights for that)

**Comment by dblock** [comment]
Thanks for doing this @rursprung!

**Comment by dblock** [comment]
@rursprung can you please manually backport it to 2.x so we can ship it sooner than later?

**Comment by rursprung** [comment]
> @rursprung can you please manually backport it to 2.x so we can ship it sooner than later?

done. thanks!
